### PR TITLE
perf(qwen36): faithful weight-stationary grouped MoE prefill for Qwen3.6-35B-A3B

### PR DESCRIPTION
**Pull request by 9876543210-tc-0123456789**
## Summary

**Faster, faithful replacement of the merged #530 grouped-MoE prefill** for Qwen3.6-35B-A3B. #530 landed the batched grouped-by-expert MoE prefill as *"the correctness-first version: a plain shared-memory-tiled bf16 grouped GEMM"* and explicitly left the GEMM inner loop as future work. This PR swaps in a **weight-stationary** expert FFN: each Q4_K/Q5_K expert weight is dequanted into shared memory **once per row-tile** (byte-faithful per-64-subblock scale+min affine, matching the decode MoE FFN) and reused across every token routed to it, instead of re-streaming per grouped-GEMM tile. Same permute/scatter structure, materially faster GEMM. Removes #530's now-superseded `moe_prefill_grouped.cu`/`moe_prefill.h` (its launchers collide by name with this path). Dense Qwythos path untouched; decode untouched.

## Proof of speedup

> Both `main` (current frontier, includes #530) and this PR built from source and benched on the same RTX 5090.

- [x] Tested on **RTX 5090** (`sm_120`)

**Prefill pp tok/s** (Qwen3.6-35B-A3B · `qwen3_gguf_bench <gguf> 128 <ctx>`), measured **vs the current frontier (main w/ #530)**:

| | prefill pp tok/s |
|---|--:|
| before prefill (main = frontier, #530) | 1115.92 |
| after prefill (this PR) | 2799.05 |

| ctx | frontier (main+#530) | this PR | marginal |
|---|--:|--:|--:|
| 4096 | 1115.92 | 2799.05 | +150.8% (2.5×) |
| 32768 | 1461.62 | 2328.20 | +59.3% |

**Correctness** (eval's long-context probe: batched prefix via `cache_prefix` + teacher-forced tail): this PR **top-1 0.9417 / PPL 1.487** — equal top-1 to the frontier (0.9417) and slightly lower PPL, comfortably over the 0.90 gate. `SPARKINFER_PREFILL_MOE=0` restores the token-loop baseline for A/B on the same binary.

```text
# qwen3_gguf_bench 128, same box, fresh source builds:
frontier (main+#530):  4k 1115.92 pp · 32k 1461.62 pp
this PR:               4k 2799.05 pp · 32k 2328.20 pp
# correctness (SPARKINFER_PREFIX_CACHE_LEN=4096 batched prefix, teacher-forced tail):
frontier: ARGMATCH 97/103 0.9417 · PPL 1.571
this PR:  ARGMATCH 97/103 0.9417 · PPL 1.487
```

### Implementation

- `kernels/csrc/cuda/moe/moe_prefill_grouped_i8.cu` (+ `moe_prefill_grouped.h`): batched router-logits, block-aligned per-expert permute with a device-built (expert → row-tile) schedule, the **weight-stationary** grouped gate/up + down kernels (fused per-subblock K-quant dequant into shared memory, reused across the tile's rows), weighted scatter-back, gated shared-expert finalize.
- Removes `moe_prefill_grouped.cu` / `moe_prefill.h` (#530's slower GEMM, superseded here).
- `runtime/src/models/qwen35_prefill.cpp` / `qwen35.cpp`: MoE FFN branch + gate, composed with #531's dense-FFN token-chunking and 128k cap already in `main`.

### CODE REVIEW

**Comment by skyrocket2026** [comment]
## ⏸ Merge conflict — rebase before eval

This branch **conflicts with `main`**, so the RTX 5090 eval is skipped until you rebase.

Please **rebase onto `main`** and push — the bot re-evaluates on the next poll once the PR is cleanly mergeable.

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
## ❌ sparkinfer auto-eval — `efe3c76`

| metric | value |
|---|---|
| **label** | `eval:REJECT` |
| Qwen3.5 score | `eval-qwen35:REJECT` (fail) |
| Qwen3.6 score | `eval-qwen36:REJECT` (fail) |
| Qwen3.5 vs same-box main | 293.87 tok/s → +0.0% (+0.1) |
| Qwen3.5 scored decode (128 ctx · 128-context) | 293.97 tok/s |
| Qwen3.5 scored prefill (131072 ctx · 128k-context) | 6887.32 pp tok/s · `eval-prefill:none` |
| Qwen3.5 correctness | top-1 89.5% · KL 0.0385 |
| Qwen3.5 128-token no-regression gate | 293.97 tok/s vs main 293.87 tok/s · pass |
| Qwen3.5 4k-context no-regression gate | 283.86 tok/s vs main 283.81 tok/s · pass |
| Qwen3.5 32k-context no-regression gate | 282.24 tok/s vs main 282.15 tok/s · pass |
| Qwen3.5 64k-context no-regression gate | 282.24 tok/s vs main 282.16 tok/s · pass |
| Qwen3.5 128k-context no-regression gate | 281.4 tok/s · pass |
| Qwen3.5 4k prefill no-regression gate | 15778.4 pp tok/s vs main 15782.96 pp tok/s · pass |
| Qwen3.5 32k prefill no-regression gate | 17341.49 pp tok/s vs main 17342.48 pp tok/s · pass |
| Qwen3.5 64k prefill no-regression gate | 17395.48 pp tok/s vs main 17396.81 pp tok/s · pass |
| Qwen3.5 128k prefill no-regression gate | 6887.32 pp tok/s vs main 6886.77 pp tok/s · pass |
| Qwen3.6 vs same-box main | 450.53 tok/s → +0.1% (+0.4) |
| Qwen3.6 scored decode (4096 ctx · 4k-context) | 450.95 tok/s |
| Qwen3.6 scored prefill (512 ctx · 512-context) | 2041.36 pp tok/s · `eval-prefill:XL` |
| Qwen3.6 correctness | top-1 89.5% · KL 0.0481 |
| Qwen3.6 128-token no-regression gate | 476.37 tok/s vs main 476.13 tok/s · pass |
| Qwen3.6 512-context no-regression gate | 470.22 tok/s vs main 469.98 tok/s · pass |
| Qwen3.6 4k-context no-regression gate | 450.95 tok/s vs main 450.53 tok/s · pass |
| Qwen3.6 16k-context no-regression gate | 434.07 tok/s vs main 434.05 tok/s · pass |
| Qwen3.6 32k-context no-regression gate | 406.64 tok/s vs main 406.52 tok/s · pass |
| Qwen3.6 128 prefill no-regression gate | 0.0 pp tok/s · pass |
| Qwen3.6 512 prefill no-regression gate | 2041.36 pp tok/s vs main 348.3 pp tok/s · pass |
| Qwen3.6 4k prefill no-regression gate | 2591.92 pp tok/s vs main 985.72 pp tok/s · pass |
| Qwen3.6 16k prefill no-regression gate | 2581.55 pp tok/s vs main 1231.03 pp tok/s · pass |
| Qwen3.6 32k prefill no-regression gate | 2224.13 pp tok/s vs main 1282.58 pp tok/s · pass |
| **Qwen3.5 optimize** | `eval:REJECT` · 293.97 tok/s · fail |
| Qwen3.5 optimize — Qwen3.6-35B-A3B guard accuracy | top-1 89.5% · KL 0.0481 · **FAIL** |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 128 | 293.97 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 4k | 283.86 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 32k | 282.24 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 64k | 282.24 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 128k | 281.4 tok/s · pass |
| **Qwen3.6 optimize** | `eval:REJECT` · 450.95 tok/s · fail |
| Qwen3.6 optimize — Qwythos-9B (Q4_K_M) guard accuracy | top-1 89.5% · KL 0.0385 · **FAIL** |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 128 | 476.37 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 512 | 470.22 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 4k | 450.95 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 16k | 434.07 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 32k | 406.64 tok/s · pass |

**Rejected** — no-regression guard: Qwen3.6-35B-A3B accuracy broke (top1=0.8952, kl=0.0481).

_RTX 5090 (sm_120) · 128/512/4k/16k/32k guarded · Qwen3.5 prefill at 4k/32k/64k/128k · Qwen3.6 prefill at 128/512/4k/16k/32k · scored vs same-box main · built from source · correctness vs llama.cpp. Automated — **not merged**; merge manually after review._

**Comment by skyrocket2026** [comment]
## ⏸ Merge conflict — rebase before eval

This branch **conflicts with `main`**, so the RTX 5090 eval is skipped until you rebase.

Please **rebase onto `main`** and push — the bot re-evaluates on the next poll once the PR is cleanly mergeable.

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by skyrocket2026** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).
